### PR TITLE
Allows keeping values for font-size/family when pasting

### LINKED ISSUE #2278: Handling variations of font-family notations and partial matches
When working on the initial implementation of the font family feature @jodator implemented a regexp-based solution which made the editor accept content like:

```html
<span style="font-family: &quot;Ba ba&quot;, 'Bu bu'">
<span style='font-family: "Bo bo"'>
```

Such content cannot be produced by the editor but it may pasted from outside of the editor.

During the review we decided to ignore such cases for now. The reasoning is that the editor should only be able to handle data which it produced for now and the whole spectrum of mismatched font name values is much wider:

(source: https://github.com/ckeditor/ckeditor5-font/pull/5#issuecomment-357954222)

> We only need to care about correctly loading data created with the editor. There's much more than quotes than can go wrong here and it's not yet time to worry about these situations. Some examples:
>
> * `font` was used instead of `font-family`,
> * some characters were encoded in `font-family`'s value,
> * there's only a partial match for a font family name – e.g. only the first name was used.
>
> All these cases may happen with a pasted content and a complete solution for them would be extremely complex. We've seen that in CKEditor 4 already where a bit too aggressive normalisation was actually breaking stuff rather than helping.

We need to understand which cases we should handle. Also, the "how" and "where" questions need to be answered.

### LINKED ISSUE #6165: Font styles are not honored when pasting from MS Word
## 📝 Provide detailed reproduction steps (if any)

1. Add, to `article.js` plugins: PasteFromOffice, Font and Highlight (last one just in case).
2. Configure CKEditor font as shown in attached `article.js` file. The configuration used should exclude possibility of something not being configured - [article.zip](https://github.com/ckeditor/ckeditor5/files/4133958/article.zip).
3. Paste contents of attached MS Word file - [testfile.zip](https://github.com/ckeditor/ckeditor5/files/4133960/testfile.zip)
 

### ✔️ Expected result

When pasting from MS Word file, font styling should be preserved.

### ❌ Actual result

No styling gets preserved.

issue2

## 📃 Other details

* Browser: Any
* OS: Any
* CKEditor version: Latest
* Installed CKEditor plugins: PasteFromOffice, Font, Highlight

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced an option for both (`FontSize` and `FontFamily`) plugins that allow keeping font-family and font-size values when pasting a text. Closes ckeditor/ckeditor5#6165. Closes ckeditor/ckeditor5#2278.

---

### Additional information

TODO: Enable the features in "Paste from ..." demos.

### CODE REVIEW

**Comment by pomek** [comment]
@Reinmar, what should we do if a user used this new option in the `FontSize` plugin in an editor with presets that were using class names? I wrote some code that covers this case but I am not sure whether it is correct (according to https://github.com/ckeditor/ckeditor5/issues/6165#issuecomment-595766519).

**Comment by pomek** [comment]
A few tests fail, trying to fix that.

File: docs/features/font.md [hunk 1]
```diff
@@ -47,6 +47,25 @@ ClassicEditor
...
+By default, all not specified values of `font-family` are removing during the conversion. You can disable this behaviour using `disableValueMatching` option.
+
+```js
+ClassicEditor
+	.create( document.querySelector( '#editor' ), {
+		fontFamily: {
+			options: [
+				// ...
+			],
+            disableValueMatching: true
```
Comment (Reinmar) [inline docs/features/font.md]:
```suggestion
			disableValueMatching: true
```

File: src/fontfamily/fontfamilyediting.js [hunk 2]
```diff
@@ -49,7 +49,8 @@ export default class FontFamilyEditing extends Plugin {
 				'Times New Roman, Times, serif',
 				'Trebuchet MS, Helvetica, sans-serif',
 				'Verdana, Geneva, sans-serif'
-			]
+			],
+			disableValueMatching: false
```
Comment (Reinmar) [inline src/fontfamily/fontfamilyediting.js]: TODO find a better name.

File: src/fontfamily/fontfamilyediting.js [hunk 3]
```diff
@@ -71,8 +72,42 @@ export default class FontFamilyEditing extends Plugin {
...
+	 *
+	 * @private
+	 */
+	_prepareAnyValueConverters() {
+		const editor = this.editor;
+
+		editor.conversion.for( 'downcast' ).attributeToElement( {
+			model: FONT_FAMILY,
+			view: ( attributeValue, writer ) => {
+				return writer.createAttributeElement( 'span', { style: 'font-family:' + attributeValue }, { priority: 7 } );
```
Comment (Reinmar) [inline src/fontfamily/fontfamilyediting.js:96]: Why priority 7?

Comment (Reinmar) [reply src/fontfamily/fontfamilyediting.js:96]: OK, it was needed to correctly sort the font color and font family and font size styles so they render in a visually correct way. But we need a test for that. Most likely, such a test exist for the previous converter, so it's a matter of copying it.

File: src/fontsize/fontsizeediting.js [hunk 4]
```diff
@@ -48,31 +48,92 @@ export default class FontSizeEditing extends Plugin {
...
-			isFormatting: true,
-			copyOnEnter: true
+		editor.conversion.for( 'downcast' ).attributeToElement( {
+			model: FONT_SIZE,
+			view: ( attributeValue, writer ) => {
+				if ( !attributeValue ) {
+					return;
+				}
+
+				if ( typeof attributeValue == 'number' ) {
```
Comment (Reinmar) [inline src/fontsize/fontsizeediting.js]: You can probably remove this.

Comment (Reinmar) [inline src/fontsize/fontsizeediting.js]: You can probably remove this.

File: src/fontsize/fontsizeediting.js [hunk 5]
```diff
@@ -48,31 +48,92 @@ export default class FontSizeEditing extends Plugin {
...
+			model: {
+				key: FONT_SIZE,
+				value: viewElement => {
+					const fontSize = viewElement.getStyle( 'font-size' );
+
+					if ( fontSize ) {
+						return fontSize;
+					}
+
+					for ( const className of viewElement.getClassNames() ) {
```
Comment (Reinmar) [inline src/fontsize/fontsizeediting.js]: I wouldn't do this. In some sense, it bring backwards compat with already created content, but that content could have used different classes anyway, so it's not really.

File: docs/features/font.md [hunk 6]
```diff
@@ -47,6 +47,25 @@ ClassicEditor
 
 {@snippet features/custom-font-family-options}
 
+### Prevent removing non-specified values
+
+By default, all not specified values of `font-family` are removing during the conversion. You can disable this behaviour using `disableValueMatching` option.
```
Comment (Reinmar) [inline docs/features/font.md]:
```suggestion
By default, all not specified values of `font-family` are removing during the conversion. You can disable this behaviour using `disableValueMatching` option.
```

**Pull request merged**